### ISSUE TITLE
Error while creating digest for application

### ISSUE THREAD

**Issue by ifbarrera** [Neiva07/review-digest-creator#1]
Hi Lucas! 👋  Thanks for putting together this project, we really enjoyed reading through the code and getting it up and running.

I wanted to flag an error we came across, hopefully it's something easy to fix! It would be really awesome  to see this working end to end.

Here's what I did:

1. Cloned the project ✅
2. Installed mongodb on my macOS machine using:

```
brew tap mongodb/brew
brew install mongodb-community@4.0
```

3. Started up mongodb:

```
brew services start mongodb-community@4.0
```

4. I then updated the `CreateApplicationDigestRequest` in `applications.go` to look like this:

```
	request := &CreateApplicationDigestRequest{
		SleepTime:     1,
		StartDigestAt: time.Now(),
		PageSize:      10,
	}
```

5. Created a `.env` file by doing `touch .env` in the root project directory, then added `DIGEST_PATH="."` to the .env file.
6. Started the app with `go run  .` and then hit the Create App Digest endpoint with `447188370` as the appId.

The app started up and set up the application correctly, looks like.

```
2022/01/27 20:53:03 Creating Application Digest for application 447188370
2022/01/27 20:53:03 Sucessfully inserted application 447188370 into the database
2022/01/27 20:53:03 Starting execution of job of application 447188370
2022/01/27 20:53:03 Application 447188370 scheduled successfully. Next run will be in:
2022/01/27 20:53:03 2022-01-28 02:53:03.042401 +0000 UTC
2022/01/27 20:53:03 Found last review Added &{447188370 8286004356 Locked account 2022-01-26 19:51:14 +0000 UTC 1 🐾IG-_ccm_tyler I been on Snapchat since it was created I never expected anything like this to happen … smh false}
```
Then it fetched the reviews and seemingly stored them properly. It was when it tried to create the digest that it errored out:

```
2022/01/27 21:53:03 Found last review Added &{447188370 8286004356 Locked account 2022-01-26 19:51:14 +0000 UTC 1 🐾IG-_ccm_tyler I been on Snapchat since it was created I never expected anything like this to happen … smh false}
2022/01/27 21:53:03 Retrieving Apple Reviews for appicationId 447188370
2022/01/27 21:53:03 Querying for https://itunes.apple.com/us/rss/customerreviews/id=447188370/sortBy=mostRecent/page=1/json url
2022/01/27 21:53:03 200 OK
2022/01/27 21:53:03 Found 15 reviews not saved from application 447188370
2022/01/27 21:53:03 Error while creating digest for application 447188370
2022/01/27 21:53:03 No reviews to digest for application 447188370
```

It seems to think there are no reviews to digest, even though there should be.

Let me know if I can provide any more info! 🙏

**Comment by Neiva07** [comment]
Hi Isabel! Thank you for reporting it!

Looking over the logs sound like the error is being created over `digest.go` file in the snippet below:

```
if len(reviews) == 0 {
		errorMessage := fmt.Sprintf("No reviews to digest for application %s", application.ApplicationId)
		return errors.New(errorMessage)
}
```

So this let me believe that something happened in the query step to have no reviews pulled from the database.

Since you changed the application `sleepTime` for 1 hour this trigger 2 side effects:

1. The application will create reports every hour.
2. The application will only digest non previously digested reviews AND will only digest reviews created in the last hour.

The second item is represented in the snippet below from the `cronJob.go`:

```
rangeTime := time.Now().Add(time.Hour * (-time.Duration(application.SleepTime)))

reviews := persistence.RetrieveNotDigestedReviews(application.ApplicationId, int64(application.PageSize), rangeTime)
```

As the requirements of the challenge suggested, should only be added reviews never digested AND from the last 24 hours. Since I parameterized the 24h to be more flexible (1 hour in your choice), I decided to also follow the requirement of only search for reviews in the last time window (1h).

Hope this clarify 🙂

Please, feel free to ask anything else in case you find another bug, odd behavior or any other question.

**Comment by ifbarrera** [comment]
Ahhh okay thanks for clarifying, that's interesting! So you're saying that the reason that there are no reviews to digest is because the new reviews that were fetched did not have a `postedAt` value that fell within the last hour.

I can see why you the assignment might be interpreted that way, but from a product perspective, the digest shouldn't care _when_ the review was posted - what it cares about is whether that individual review has been part of a "digest" in the past. 

So for the 1 hour example, if we fetch reviews with ids 1,2, and 3 at 12pm, and then, at 1pm, we refetch reviews, and receive reviews 1, 2, 3, and 4, I would expect a digest to be created with review #4, since that one is "new" since the last time we fetched and showed a digest. The reason the `postedAt` date might be misleading is because the iTunes API doesn't synchronously process reviews as they come in - so if I post a review today, the API might not return it until a few days from now.

**Comment by Neiva07** [comment]
Interesting! So what you're saying is actually the date/time of post shouldn't be relevant at all when it comes to digest reviews, right? The only relevant factor would be if the review was already digested before.

I think this would be pretty simple to change, the only actual change would be in the query that will no longer depend upon `postedAt` and I might need to remove this field from the index as well.

 I can make a new commit as soon as you confirm that, what do you think? :)

**Comment by ifbarrera** [comment]
Yep exactly! We want to make sure the digest shows any new reviews that are being returned from the API since the last time we showed a digest.
